File: Python/IslandPerimeter.py
from collections import deque

def islandPerimeter(grid):

    ROWS,COLS = len(grid), len(grid[0])

    queue = deque()

    perimeter = 0

    for i in range(ROWS):
        for j in range(COLS):
            if grid[i][j] == 1:
                perimeter+=4
                queue.append((i,j))


    while queue:
        for i in range(len(queue)):
            r,c = queue.popleft()

            directions = [[1,0],[-1,0],[0,1],[0,-1]]
            for x,y in directions:
                dr = x+r
                dc = y+c

                if dr>=0 and dc>=0 and dr<ROWS and dc<COLS and grid[dr][dc] == 1:
                    perimeter-=1

    return perimeter

File: Javascript/IslandPerimeter.js
var islandPerimeter = function(grid){
    const [ROWS,COLS] = [grid.length,grid[0].length]
    const queue = []
    let perimeter = 0

    for (let i =0;i<ROWS;i++){
        for(let j=0;j<COLS;j++){
            if(grid[i][j] === 1){
                perimeter += 4
                queue.push([i,j])
            }
        }
    }

    while(queue.length>0){
        const queueLength = queue.length
        for(let i=0;i<queueLength;i++){
            const [r,c] = queue.pop()

            const directions = [[1,0],[-1,0],[0,-1],[0,1]]
            for (const [x,y] of directions){
                const dr = x + r
                const dc = y + c

                if (dr>=0 && dc>=0 && dr<ROWS && dc<COLS && grid[dr][dc] === 1){
                    perimeter-=1
                }
            }
        }
    }

    return perimeter
}